src/base/http/requestparser.cpp

```diff
@@ -30,314 +31,307 @@
+                qDebug() << Q_FUNC_INFO << "incomplete request";
+                return {ParseStatus::Incomplete, Request(), 0};
+            }
+            if (httpBody.length() != contentLength) {
+                qDebug() << Q_FUNC_INFO << "content length mismatch";
+                return {ParseStatus::BadRequest, Request(), 0};
+            }
 
-        if ((contentLength > 0) && !parseContent(content)) {
-            qWarning() << Q_FUNC_INFO << "message parsing error";
-            return BadRequest;
+            if (!parsePostMessage(httpBody)) {
+                qWarning() << Q_FUNC_INFO << "message body parsing error";
+                return {ParseStatus::BadRequest, Request(), 0};
+            }
         }
-    }
 
-    //  qDebug() << Q_FUNC_INFO;
-    //  qDebug() << "HTTP Request header:";
-    //  qDebug() << data.left(headerEnd) << "\n";
+        return {ParseStatus::OK, m_request, (headerLength + contentLength)};
+    }
 
-    request = m_request;
-    return NoError;
+    qWarning() << Q_FUNC_INFO << "unsupported request method: " << m_request.method;
+    return {ParseStatus::BadRequest, Request(), 0};  // TODO: SHOULD respond "501 Not Implemented"
 }
 
-bool RequestParser::parseStartingLine(const QString &line)
+bool RequestParser::parseStartLines(const QString &data)
 {
-    const QRegExp rx("^([A-Z]+)\\s+(\\S+)\\s+HTTP/\\d\\.\\d$");
-
-    if (rx.indexIn(line.trimmed()) >= 0) {
-        m_request.method = rx.cap(1);
-
-        QUrl url = QUrl::fromEncoded(rx.cap(2).toLatin1());
-        m_request.path = url.path(); // Path
-
-        // Parse GET parameters
-        QListIterator<QPair<QString, QString> > i(QUrlQuery(url).queryItems());
-        while (i.hasNext()) {
-            QPair<QString, QString> pair = i.next();
-            m_request.gets[pair.first] = pair.second;
+    // for simplicity, we don't handle malformed request which uses `LF` for newline
+    const QVector<QStringRef> lines = data.splitRef(CRLF, QString::SkipEmptyParts);
+
+    // [rfc7230] 3.2.2. Field Order
+    QStringList requestLines;
+    for (const auto &line : lines) {
+        if (line.at(0).isSpace() && !requestLines.isEmpty()) {
+            // continuation of previous line
+            requestLines.last() += line.toString();
+        }
+        else {
+            requestLines += line.toString();
         }
-
-        return true;
     }
 
-    qWarning() << Q_FUNC_INFO << "invalid http header:" << line;
-    return false;
-}
+    if (requestLines.isEmpty())
+        return false;
 
-bool RequestParser::parseHeaderLine(const QString &line, QPair<QString, QString> &out)
-{
-    int i = line.indexOf(QLatin1Char(':'));
-    if (i == -1) {
-        qWarning() << Q_FUNC_INFO << "invalid http header:" << line;
+    if (!parseRequestLine(requestLines[0]))
         return false;
+
+    for (auto i = ++(requestLines.begin()); i != requestLines.end(); ++i) {
+        if (!parseHeaderLine(*i, m_request.headers))
```

Comment:
Why parseHeaderLine can't access m_request.headers directly?

Reply:
parseHeaderLine is used 2 times with different parameters.

Reply:
Then it can be defined outside RequestParser class. Unless I'm wrong, it's independent from this class.

Reply:
>Then it can be defined outside RequestParser class. 

Move to namespace above?

Reply:
Yes.